### ISSUE TITLE
EKS ClusterLoggingArgs is incorrectly defined

### ISSUE THREAD

**Issue by aschleck** [pulumi/pulumi-aws-native#644]
### What happened?

I tried to import an EKS cluster into my Pulumi configuration that defines a logging configuration.

### Steps to reproduce

Try to define the `logging` variable in an `eks.Cluster`.

### Expected Behavior

I expected to be able to define ClusterLogging corresponding to

```
      ~ logging                 : {
          ~ clusterLogging: {
              - enabledTypes: [
              -     [0]: {
                      - type: "api"
                    }
              -     [1]: {
                      - type: "audit"
                    }
              -     [2]: {
                      - type: "authenticator"
                    }
                ]
            }
        }
```

### Actual Behavior

When running `pulumi import`, I first saw:

```
    warning: aws-native:eks:Cluster resource 'cow-cluster': property logging/clusterLogging.enabledTypes value {<nil>} has a problem: unknown property logging/clusterLogging.enabledTypes
```

and Pulumi's generated code included

```
    logging=aws_native.eks.ClusterLoggingArgs(
        cluster_logging=aws_native.eks.ClusterLoggingArgs(),
    ),
```

And you can see the API docs show it recursively defined: https://www.pulumi.com/registry/packages/aws-native/api-docs/eks/cluster/#clusterlogging

And the code confirms it: https://github.com/pulumi/pulumi-aws-native/blob/b818566c35b7d8e2d9e03998dc9cff80788d201f/sdk/python/pulumi_aws_native/eks/outputs.py#L194

So, at least following the typing, there is no way to define my logging configuration.

### Output of `pulumi about`

```
CLI
Version      3.39.1
Go Version   go1.19
Go Compiler  gc

Plugins
NAME           VERSION
aws-native     0.24.0
google-native  0.25.0
python         unknown

Host
OS       ubuntu
Version  22.04
Arch     x86_64

This project is written in python: executable='<snip>' version='3.10.2
'

Backend
Name           <snip>
URL            <snip>
User           april
Organizations

Dependencies:
NAME                  VERSION
pip                   22.0.3
pulumi-aws-native     0.24.0
pulumi-google-native  0.25.0
setuptools            60.9.3

Pulumi locates its logs in /tmp by default
warning: Failed to get information about the current stack: No current stack
```

### Additional context

It's really unbelievable how much better using a real language is compared to HCL, thank you for this wonderful software

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by aschleck** [comment]
Also, I tried adding an `ignore_changes=["logging"]` to proceed with an import directive defined in `ResourceOptions` but Pulumi fails to import:

```
    error: aws-native:eks:Cluster resource 'cow-cluster': property logging/clusterLogging.enabledTypes value {<nil>} has a problem: unknown property logging/clusterLogging.enabledTypes
    error: Preview failed: one or more inputs failed to validate
```

**Comment by danielrbradley** [comment]
Thanks for logging this @aschleck 

It does look a little odd that the ClusterLogging just refers to its own type. [According to the specifications](https://github.com/pulumi/pulumi-aws-native/blob/master/aws-cloudformation-schema/aws-eks-cluster.json#L85-L94), this should just have a single property called `EnabledTypes`.

Schema for the ClusterLogging type:

```json
         "aws-native:eks:ClusterLogging": {
            "description": "Enable exporting the Kubernetes control plane logs for your cluster to CloudWatch Logs based on log types. By default, cluster control plane logs aren't exported to CloudWatch Logs.",
            "properties": {
                "clusterLogging": {
                    "$ref": "#/types/aws-native:eks:ClusterLogging",
                    "description": "The cluster control plane logging configuration for your cluster. ",
                    "language": {
                        "csharp": {
                            "name": "ClusterLoggingValue"
                        }
                    }
                }
            },
            "type": "object"
        },
```

**Comment by aschleck** [comment]
Okay I spent some time debugging this, haven't wrapped my mind totally around the code but here's a fuzzy stab at what's happening:

Inside of https://github.com/pulumi/pulumi-aws-native/blob/74e59199a4f207c3508998297b0793f49fc74123/provider/pkg/schema/gen.go#L821 the properties get the type of their parent prepended to their names and that definition is cached. So eg OutpostConfig becomes ClusterOutpostConfig. This is mostly great, except when it isn't. Cluster has a Logging property, which becomes ClusterLogging. Which means when we try to resolve the type of the Logging property (which is typed ClusterLogging), it uses the cached the definition from its parent. This is how it ends up recursive.

As far as fixing this, I suppose maybe one option is to put some sort of special character in the same way there's a special character in front of the generated Input types. What do you folks think?

**Comment by aschleck** [comment]
I'm actually not sure how to fix this. It's easy enough to make a patch that generates the correct types.

```
diff --git a/provider/pkg/schema/gen.go b/provider/pkg/schema/gen.go
index c6548ef37..5acbef838 100644
--- a/provider/pkg/schema/gen.go
+++ b/provider/pkg/schema/gen.go
@@ -824,9 +824,9 @@ func (ctx *context) propertyTypeSpec(parentName string, propSchema *jsschema.Sch
                schemaName := strings.TrimPrefix(propSchema.Reference, "#/definitions/")
                typName := schemaName
                if !strings.HasPrefix(schemaName, ctx.resourceName) {
-                       typName = fmt.Sprintf("%s%s", ctx.resourceName, schemaName)
+                       typName = fmt.Sprintf("%s_%s", ctx.resourceName, schemaName)
                }
                tok := fmt.Sprintf("%s:%s:%s", packageName, ctx.mod, typName)
```

But that would be a major breaking change. All the types like Logging get prefixed with resource names like Cluster, so when there actually is a non-resource type named ClusterLogging it will conflict. Maybe the best thing to do is detect the collision and name non-resource one "InnerClusterLogging" or something?

@lblackstone any thoughts?

**Comment by danielrbradley** [comment]
> it uses the cached the definition from its parent. This is how it ends up recursive.

Thank you for your investigation here - that's very helpful. Indeed I don't think we can add the change to the type name as it will be a breaking change - though this might be worth considering at the point of the next major version to simplify things.

### Notes on possible solutions

Initial options:
- Add additional checking beyond the type name existing to resolve these cases - not clear how to differentiate the name in a stable way when there is a conflict though.
- Add a list of special-cases to override the naming in specific scenarios. This could just be in place until we can introduce the breaking change.

Note: If using a manually curated list, it would be good to find a way to report all potential clashes - possibly even failing the build until they've been resolved manually.

**Comment by aschleck** [comment]
I can't quite recall right now, but I think I dumped every collision and ClusterLogging is the only type that I saw with this issue. Special-casing makes the most sense to me, but I of course defer to your sense since you know more than I do.
